agent/src/util/device_plugin_service.rs

```diff
@@ -403,48 +455,62 @@ impl InstanceDevicePlugin {
-            Err(Status::new(
-                Code::Unknown,
-                "Requested device already in use",
-            ))
+    instance_name: &str,
+    instance_namespace: &str,
+    capacity: &i32,
+    kube_interface: Arc<impl KubeInterface>,
+) -> Vec<(String, DeviceUsageStatus)> {
+    let mut device_usage_states = Vec::new();
+    match kube_interface
+        .find_instance(instance_name, instance_namespace)
+        .await
+    {
+        Ok(kube_akri_instance) => {
+            for (device_name, device_usage_string) in kube_akri_instance.spec.device_usage {
+                let device_usage_status = match DeviceUsage::from_str(&device_usage_string) {
+                    Ok(device_usage) => get_device_usage_state(&device_usage, node_name),
+                    Err(_) => DeviceUsageStatus::Unknown,
+                };
+                device_usage_states.push((device_name.clone(), device_usage_status));
+            }
+            device_usage_states
         }
-    } else {
-        // No corresponding id found
-        trace!(
-            "slot_available_to_reserve - could not find {} id in device_usage",
-            device_usage_id
-        );
-        Err(Status::new(
-            Code::Unknown,
-            "Could not find device usage slot",
-        ))
+        Err(_) => (0..*capacity)
+            .map(|x| {
+                (
+                    format!("{}-{}", instance_name, x),
+                    DeviceUsageStatus::Unknown,
+                )
+            })
+            .collect(),
     }
 }
 
+/// This returns device usage status of a `device_usage_id` slot for an instance on a given node
+/// # More details
+/// Cases based on the device usage value
+/// 1. DeviceUsageKind::Free ... this means that the device is available for use
+///     * (ACTION) return DeviceUsageStatus::Free
+/// 2. device_usage.usage == node_name ... this means node_name previously used device_usage
+///     * (ACTION) return previously reserved kind, DeviceUsageStatus::ReservedByConfiguration or DeviceUsageStatus::ReservedByInstance
+/// 3. device_usage.usage == (some other node) ... this means that we believe this device is in use by another node
+///     * (ACTION) return DeviceUsageStatus::ReservedByOtherNode
+fn get_device_usage_state(device_usage: &DeviceUsage, node_name: &str) -> DeviceUsageStatus {
+    let device_usage_state = match device_usage.get_kind() {
+        DeviceUsageKind::Free => DeviceUsageStatus::Free,
+        DeviceUsageKind::Configuration(vdev_id) => {
+            DeviceUsageStatus::ReservedByConfiguration(vdev_id)
+        }
+        DeviceUsageKind::Instance => DeviceUsageStatus::ReservedByInstance,
```

Comment:
what happens if Unknown is found?

Reply:
This function return DeviceUsageStatus based on DeviceUsageKind and node name, it has sufficient information to determine the DeviceUsageStatus and never returns DeviceUsageStatus::Unkown

Reply:
sorry, my bad!  there is no DeviceUsageKind::Unknown ... only a DeviceUsageStatus::Unknown.